### ISSUE TITLE
Oracle Timestamp with Time Zone not being mapped to DateTimeOffset properly

### ISSUE THREAD

**Issue by Treyphor** [StackExchange/Dapper#1039]
I have a created date column with type of Timestamp with Time Zone in oracle. My C# model property type is a DateTimeOffset. It sets the value, but it loses the offset portion.

I tried writing my own custom type handler, but the object value that gets passed to the parse method is already a DateTime object, and has already lost the offset value of the data at that point.

The issue is reader => handler.Parse(type, reader.GetValue(startBound));

reader.GetValue(startBound) returns a DateTime at that point.

We need to be able to change it to reader.GetDateTimeOffset() or have some sort of lower level extension point.

**Comment by NikitaEmelyanov** [comment]
Any update on this?

**Comment by Treyphor** [comment]
@NikitaEmelyanov forgot about this, but since I never heard anything I can share what I did to make this work. First, I made this custom mapper:
`public class DateTimeOffsetTypeHandler : SqlMapper.TypeHandler<DateTimeOffset>
	{
		public override void SetValue(IDbDataParameter parameter, DateTimeOffset value)
		{
			parameter.Value = value.ToOffset(TimeSpan.Zero).ToString("dd-MMM-yyyy h:mm:ss tt K");
			parameter.DbType = DbType.String;
		}

		public override DateTimeOffset Parse(object value)
		{
			switch (value)
			{
				case DateTime time:
					return new DateTimeOffset(DateTime.SpecifyKind(time, DateTimeKind.Utc), TimeSpan.Zero);
				case DateTimeOffset dto:
					return dto;
				default:
					throw new InvalidOperationException("Must be DateTime or DateTimeOffset object to be mapped.");
			}
		}
	}`

Then in my autofac module just before I register my connections I call this method:
`void AddCustomDapperMappers()
		{
			SqlMapper.RemoveTypeMap(typeof(DateTimeOffset?));
			SqlMapper.RemoveTypeMap(typeof(DateTimeOffset));
			SqlMapper.AddTypeHandler(typeof(DateTimeOffset), new DateTimeOffsetTypeHandler());
		}`
But you could call that in your startup class or wherever you need to. Hope this helps.

**Comment by NikitaEmelyanov** [comment]
Thanks, @Treyphor. I think it helps in some scenarios. 

Unfortunately, that isn't our case. Currently, it is impossible for us to change the way data is written to the database. Thus, it is important for us to get a “true” offset from the database, which may have been brought by another application.
So we ended up converting the date to a formatted string as suggested here: https://stackoverflow.com/questions/36021582/how-to-read-time-offsets-from-oracle-using-dapper

**Comment by leosbrf** [comment]
I don't think this issue comes from Dapper. In my tests, I noticed that the problem was the driver used for data access.

1. When using the Devart.Data.Oracle (v9.14.*), the issue reported here appears.
1. When using the Oracle.ManagedDataAccess.Core (v3.21.3), the column is mapped correctly.

**Comment by MykhailoSalivonov** [comment]
I have the same issue using Oracle.ManagedDataAccess 21.5.0. In the Parse method, I always have **case DateTime time**
